store/sql_job_store.go

```diff
@@ -0,0 +1,327 @@
+		if sqlResult, err := jss.GetMaster().Exec(
+			`UPDATE
+				Jobs
+			SET
+				LastActivityAt = :LastActivityAt,
+				Status = :Status,
+				Progress = :Progress,
+				Data = :Data
+			WHERE
+				Id = :Id
+			AND
+				Status = :OldStatus`,
+			map[string]interface{}{
+				"Id":             job.Id,
+				"OldStatus":      currentStatus,
+				"LastActivityAt": model.GetMillis(),
+				"Status":         job.Status,
+				"Data":           job.DataToJson(),
+				"Progress":       job.Progress,
+			}); err != nil {
+			result.Err = model.NewLocAppError("SqlJobStore.UpdateOptimistically",
+				"store.sql_job.update.app_error", nil, "id="+job.Id+", "+err.Error())
+		} else {
+			rows, err := sqlResult.RowsAffected()
+
+			if err != nil {
+				result.Err = model.NewLocAppError("SqlJobStore.UpdateStatus",
+					"store.sql_job.update.app_error", nil, "id="+job.Id+", "+err.Error())
+			} else {
+				if rows == 1 {
+					result.Data = true
+				} else {
+					result.Data = false
+				}
+			}
+		}
+
+		storeChannel <- result
+		close(storeChannel)
+	}()
+
+	return storeChannel
+}
+
+func (jss SqlJobStore) UpdateStatus(id string, status string) StoreChannel {
+	storeChannel := make(StoreChannel, 1)
+
+	go func() {
+		result := StoreResult{}
+
+		job := &model.Job{
+			Id:             id,
+			Status:         status,
+			LastActivityAt: model.GetMillis(),
+		}
+
+		if _, err := jss.GetMaster().UpdateColumns(func(col *gorp.ColumnMap) bool {
+			return col.ColumnName == "Status" || col.ColumnName == "LastActivityAt"
+		}, job); err != nil {
+			result.Err = model.NewLocAppError("SqlJobStore.UpdateStatus",
+				"store.sql_job.update.app_error", nil, "id="+id+", "+err.Error())
+		}
+
+		if result.Err == nil {
+			result.Data = job
+		}
+
+		storeChannel <- result
+		close(storeChannel)
+	}()
+
+	return storeChannel
+}
+
+func (jss SqlJobStore) UpdateStatusOptimistically(id string, currentStatus string, newStatus string) StoreChannel {
+	storeChannel := make(StoreChannel, 1)
+
+	go func() {
+		result := StoreResult{}
+
+		var startAtClause string
+		if newStatus == model.JOB_STATUS_IN_PROGRESS {
+			startAtClause = `StartAt = :StartAt,`
+		}
+
+		if sqlResult, err := jss.GetMaster().Exec(
+			`UPDATE
+				Jobs
+			SET `+startAtClause+`
+				Status = :NewStatus,
+				LastActivityAt = :LastActivityAt
+			WHERE
+				Id = :Id
+			AND
+				Status = :OldStatus`, map[string]interface{}{"Id": id, "OldStatus": currentStatus, "NewStatus": newStatus, "StartAt": model.GetMillis(), "LastActivityAt": model.GetMillis()}); err != nil {
+			result.Err = model.NewLocAppError("SqlJobStore.UpdateStatus",
+				"store.sql_job.update.app_error", nil, "id="+id+", "+err.Error())
+		} else {
+			rows, err := sqlResult.RowsAffected()
+
+			if err != nil {
+				result.Err = model.NewLocAppError("SqlJobStore.UpdateStatus",
+					"store.sql_job.update.app_error", nil, "id="+id+", "+err.Error())
+			} else {
+				if rows == 1 {
+					result.Data = true
+				} else {
+					result.Data = false
+				}
```

Comment:
That's true, but as you point out, this pattern is commonly used elsewhere in this codebase, so I think I'll stick with it for now.